Feature: Auto-update client

mohamed-essam opened the pull request:
## Describe your changes

## Issue ticket number and link

## Stack

<!-- branch-stack -->

### Checklist
- [ ] Is it a bug fix
- [ ] Is a typo/documentation fix
- [ ] Is a feature enhancement
- [ ] It is a refactor
- [ ] Created tests that fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary
- [x] Documentation is **not needed**

> By submitting this pull request, you confirm that you have read and agree to the terms of the [Contributor License Agreement](https://github.com/netbirdio/netbird/blob/main/CONTRIBUTOR_LICENSE_AGREEMENT.md).


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Configurable auto-update: clients can auto-update to latest, a specific version, or be disabled; initial auto-update applied at startup.
  * Update progress UI: displays live progress during client upgrades.
  * Platform update support: macOS, Windows, Linux (and placeholders for other targets).

* **Server & Sync**
  * Management surfaces auto-update settings in sync payloads and account APIs; emits events when account auto-update version changes.

* **Tests**
  * Added coverage for update manager and version decision logic

* **Documentation**
  * auto_update_version exposed in account settings API

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

robertgro commented:
Related

https://github.com/netbirdio/netbird/issues/1793

https://github.com/netbirdio/netbird/issues/4019

mohamed-essam commented on client/internal/updatemanager/update_windows.go:
```diff
@@ -0,0 +1,51 @@
+
+func (u *UpdateManager) triggerUpdate(targetVersion string) error {
+	k, err := registry.OpenKey(registry.LOCAL_MACHINE, `SOFTWARE\Microsoft\Windows\CurrentVersion\App Paths\Netbird`, registry.QUERY_VALUE)
+	if err != nil && strings.Contains(err.Error(), "system cannot find the file specified") {
```
Makes sense, what would you suggest instead? Simply matching on `err != nil`?
I'm worried any random error could cause an EXE installation to be mistaken for MSI

robertgro replied:
So a possible solution may be a nested one:
```
	const uninstallKeyPath = `SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall\Netbird`
	const uninstallKeyPath2 = `SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\Netbird`

	k, err := registry.OpenKey(registry.LOCAL_MACHINE, uninstallKeyPath, registry.READ)
	if err != nil {
		kv, err := registry.OpenKey(registry.LOCAL_MACHINE, uninstallKeyPath2, registry.READ)
		if err != nil {
			return err
		}
		defer kv.Close()
                return kv
	}
	defer k.Close()
        return k
```

Why do we need to double check here? Obviously when a 32-bit process tries to access certain registry keys, Windows redirects it to the WOW6432Node. Sometimes there can even be redirecting issues, so this extra verification is required.

In case Netbird has been installed via .msi:
- this always throws an error

In case Netbird has been installed via .exe:
- no error thrown

mohamed-essam replied:
Thank you @robertgro for the suggestion :pray:

lixmal commented on client/internal/updatemanager/update_windows.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+const (
+	msiDownloadURL = "https://github.com/netbirdio/netbird/releases/download/v%s/netbird_installer_%s_windows_amd64.msi"
```
There will be https://github.com/netbirdio/netbird/pull/4206

lixmal commented on client/internal/updatemanager/update_darwin.go:
```diff
@@ -0,0 +1,111 @@
+		// netbird-ui is installed
+		upgradeArgs = append(upgradeArgs, "netbirdio/tap/netbird-ui")
+	}
+	cmd = exec.Command("sudo", upgradeArgs...)
```
Can this get stuck if sudo is configured to ask for a password?

mohamed-essam replied:
It _can_ but I don't believe I've ever seen a system where sudo is configured to ask for password from root, and as far as I understand, the NetBird daemon always runs as root, no?

lixmal reviewed (commented):
Have you put any thoughts into how we could do a rollback if the new version breaks connectivity?

mohamed-essam commented on client/internal/updatemanager/update_darwin.go:
```diff
@@ -0,0 +1,111 @@
+		return u.updateHomeBrew()
+	}
+	// Installed using pkg file
+	path, err := downloadFileToTemporaryDir(u.ctx, urlWithVersionArch(pkgDownloadURL, targetVersion))
```
It would make sense, but I believe this can be added in a later release, but I believe it'd make more sense to verify the hash from Github since it's already there (and we already know the exact version we want to upgrade to) no?

lixmal replied:
The attack vector here is a manipulated release in github (even if signed). Pulling hashes from github in this case would only help file integrity, not security.
But yes, can be a follow-up

mohamed-essam commented:
@lixmal If it's alright with you, I will mark the PR as no documentation needed for the time being as the dashboard changes are not yet done, as soon as dashboard can modify the auto upgrade version I will open a PR in docs

lixmal commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,154 @@
+	if u.version != v {
+		log.Tracef("Auto-update version set to %s", v)
+		u.version = v
+		go u.CheckForUpdates()
```
This will run once per network map, right? It should  debounce internally to avoid running multiple of these.

mohamed-essam replied:
Locked entire `CheckForUpdates` method with a mutex, if there's an update we won't need to worry about it anyway since the entire Process will shutdown

lixmal commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+func (u *UpdateManager) SetVersion(v string) {
+	if u.version != v {
```
Let's either add a mutex here or a comment on top of the method that this is synchronized externally (`syncMsgMux`)

lixmal commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,154 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	manager := &UpdateManager{
+		update:         update,
+		lastTrigger:    time.Now().Add(-10 * time.Minute),
```
This field is not thread-safe

mohamed-essam replied:
Locked entire `CheckForUpdates` method with a mutex

lixmal commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,154 @@
+			log.Tracef("Latest version not fetched yet")
+			return
+		}
+		updateVersionString = u.update.LatestAvailable.String()
```
Alright

lixmal commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,154 @@
+	currentVersionString := version.NetbirdVersion()
+	updateVersionString := u.version
+	if updateVersionString == latestVersion || updateVersionString == "" {
+		if u.update.LatestAvailable == nil {
```
Right, but that's a different lock than the one `update` uses when it accesses LatestAvailable. I don't think it should be exported in the first place

pappz commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,158 @@
+}
+
+func (u *UpdateManager) Stop() {
+	u.update.StopWatch()
```
The `u.update.StopWatch() `contains a bug. It can miss the "channel close" event because the channel is unbuffered. It leak a go thread.

mohamed-essam replied:
I don't understand your point, as I can see, since the channel is unbuffered, it will wait for the other goroutine to recieve the event in https://github.com/netbirdio/netbird/pull/4256/files#diff-9bf0202e6bc7e6197efaa70257c72730aac7b94dd106bf73ec4dce218fc58e29R62 and won't return until the other goroutine is dead, and I don't see the channel closed anywhere

pappz replied:
We have two goroutines. If you run the 2. routine at that moment when the u.checkUpdate is in progress, then the 2. thread writes nothing to the channel, just use the default path. You are right, do not have channels closed. I mean that "close" event on channels.
1.
```
select {
		case <-u.fetchDone:
			return
		case <-u.fetchTicker.C:
			if changed := u.fetchVersion(); changed {
				u.checkUpdate()
			}
		}
```
2.
```
select {
	case u.fetchDone <- struct{}{}:
	default:
	}
```

pappz commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,158 @@
+import (
+	"context"
+	"fmt"
+	"github.com/netbirdio/netbird/client/internal/peer"
```
Please follow our order rules of import (std lib, thirparty, NetBird)

mohamed-essam commented:
@pappz @lixmal I've changed the way goroutines and contexts are handled for the PR:

1. version.Update now closes the channel synchronously, so `updatemanager.Stop()` should now wait for version.Update's goroutine to exit.
2. Un-exported version.Update.latestVersion and made it so that it would send a copy of the latest version to the onUpdate handler instead.
3. Added a mutex lock around the entire update logic.
4. Added a child context with Deadline for all update logic.
5. Added a waitGroup for all update logic to ensure `updatemanager.Stop()` completely waits for updatemanager to exit gracefully.
6. Ensured context done would fast-fail update logic if context was done before it obtained mutex lock.

hakansa commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,183 @@
+}
+
+func (u *UpdateManager) Updated(latestVersion string) {
+	u.waitGroup.Add(1)
```
aggre with @pappz, current behaviour against the best practices. Instead of passing goroutine to *UpdateManager.Updated, you should increment the wg before you call it. e.g you should sth like this on line 64:

```go
go func (wg *sync.WaitGroup){
    wg.Add(1)
    defer wg.Done()
    go u.Updated(unknownVersion)
}(wg)


```

mohamed-essam replied:
I understand what to do regarding line 64's call of Updated, but I'm not sure what to do regarding version.Update's call

hakansa replied:
I'm not super familiar with the design but, why do you need a waitGroup there and a mutex it not enough? There will be only one goroutine handle updates right?

mohamed-essam replied:
This is the current setup, the issue is that version.Update runs in its own goroutine and spawns a new goroutine for each onUpdate, which causes the number of goroutines running Updated to be (theoretically) unlimited, practically it should only trigger once every release (+ once on connect for getting the latest release then)
And for the SetVersion it's triggered per Sync (but only calls Updated when the version is different) 
In practice this update logic isn't highly parallel, so everything we're talking about here is an extreme edge case

mohamed-essam commented:
@lixmal There's nothing we can do from inside the service, especially that (for example) the windows installer stops the service, then uninstalls, and installs.

I feel like this should be more in the responsibility of the installer itself, (i.e the installer should rollback its own changes on failure), I'm not sure if other installation methods are able to handle this that way, but that will require a lot of work on multiple installation methods

Although I'm sure some installation methods do in fact handle failure more gracefully (e.g apt doesn't restart the service until after installation is complete)

pappz commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,193 @@
+
+func (u *UpdateManager) Stop() {
+	u.cancel()
+	u.mutex.Lock()
```
What do you protect here? If I see well, you do not write the update just in the constructor.

pappz commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,193 @@
+		return
+	}
+	currentVersionString := version.NetbirdVersion()
+	updateVersionString := u.version
```
u.version is not protected by a lock

pappz commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,247 @@
+	u.expectedVersionMutex.Unlock()
+
+	// Resolve "latest" to actual version
+	if u.updateToLatestVersion {
```
It is outside of the protected scope. Create a temp variable for this, too.

pappz commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,247 @@
+
+	// Resolve "latest" to actual version
+	if u.updateToLatestVersion {
+		if !u.isVersionAvailable() {
```
Please use a new variable for this, because if you run it twice, the data might become deprecated during this period.

pascal-fischer commented on management/server/token_mgr.go:
```diff
@@ -210,6 +210,7 @@ func (m *TimeBasedAuthSecretsManager) pushNewTURNAndRelayTokens(ctx context.Cont
 		NetbirdConfig: &proto.NetbirdConfig{
 			Turns: turns,
 		},
+		AutoUpdateVersion: skipAutoUpdateVersion,
```
@mlsmaycon Why???

bcmmbaga commented on management/server/account.go:
```diff
@@ -471,6 +475,15 @@ func (am *DefaultAccountManager) handleGroupsPropagationSettings(ctx context.Con
 	}
 }
 
+func (am *DefaultAccountManager) handleAutoUpdateVersionSettings(ctx context.Context, oldSettings, newSettings *types.Settings, userID, accountID string) {
+	if oldSettings.AutoUpdateVersion != newSettings.AutoUpdateVersion {
+		am.StoreEvent(ctx, userID, accountID, accountID, activity.AccountAutoUpdateVersionUpdated, map[string]any{
+			"old_value": oldSettings.AutoUpdateVersion,
+			"new_value": newSettings.AutoUpdateVersion,
+		})
+	}
+}
```
I think we  could simplifying the meta to just store current version instead of both `old_value` and `new_value`. The event sequence already captures the change history, so we don't need the redundancy.
 
 ```suggestion
func (am *DefaultAccountManager) handleAutoUpdateVersionSettings(ctx context.Context, oldSettings, newSettings *types.Settings, userID, accountID string) {
	if oldSettings.AutoUpdateVersion != newSettings.AutoUpdateVersion {
		am.StoreEvent(ctx, userID, accountID, accountID, activity.AccountAutoUpdateVersionUpdated, map[string]any{
			"version": newSettings.AutoUpdateVersion,
		})
	}
}
```

yshah-aromatech commented:
Hi Team, 

Do we have an expected date for this release? 

Thanks

pappz commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,261 @@
+		return
+	}
+
+	ctx, cancel := context.WithDeadline(ctx, time.Now().Add(time.Minute))
```
WithTimeout simpler

mohamed-essam commented on client/internal/updatemanager/manager.go:
```diff
@@ -0,0 +1,261 @@
+	u.cancel()
+	if u.update != nil {
+		u.update.StopWatch()
+		u.update = nil
```
This is only referenced inside `handleUpdate` once inside a mutex, so I used this to prevent a race and also added a check so that if u.update is nil it should return
